### ISSUE TITLE
connectedcomponentswithstats Is Not Thread-Safe

### ISSUE THREAD

**Issue by miladaz** [opencv/opencv#13191]
##### System information (version)

- OpenCV =>3.3.1
- Operating System / Platform => Windows 10 - 64 Bit
- Compiler => Visual Studio 2015

##### Detailed description

connectedcomponentswithstats() is not thread-safe. It hangs sometimes (maybe 5% of the times). I'm guessing that it happens in the first few threads.

##### Steps to reproduce

This program should always return "5 5" (we have 5 threads) but sometimes it returns "5 4" (one hangs).
Something that can't be noticed easily when using the program because it triggers no errors.

```
// std::atomic<int> first = 0, last = 0;
ThreadPool pool(thread::hardware_concurrency()); // 2 threads
for (int i = 0; i < 5; i++) {
    pool.enqueue([i, binaryImg] { // Add the thread.
        // code ...
        first++;
        Mat labels, stats, centroids;
        int objCount = connectedComponentsWithStats(binaryImg, labels, stats, centroids, 8); // PROBLEM. Sometimes the following lines inside this block won't be executed.
        last++;
        // code ...
    });
}
pool.~ThreadPool(); // Join all the threads.
cout << first << ' ' << last << endl;

```

**Comment by JESUSROYETH** [comment]
Hi @miladaz 👋

I tried to reproduce this by calling `connectedComponentsWithStats` many times from a thread pool, and in my case it never hangs and never returns a wrong result. In total ~21k concurrent calls across several configurations: 8 and 16 threads, same binary image for all threads and a different random image per thread, with OpenCV's internal parallelism at its default (32 threads here) and also forced to 1 with `cv2.setNumThreads(1)`. OpenCV releases the GIL inside native calls, so the calls really run in parallel. Tested on OpenCV 4.10.0 and 5.0.0, Python 3.12.3, Ubuntu 24.04, x86-64.

```python
import threading, numpy as np, cv2

np.random.seed(42)
noise = cv2.GaussianBlur(np.random.rand(512, 512).astype(np.float32), (15, 15), 0)
binary = (noise > 0.5).astype(np.uint8) * 255
ref_n, ref_labels, ref_stats, _ = cv2.connectedComponentsWithStats(binary, 8)

finished, errors, lock = 0, [], threading.Lock()

def worker():
    global finished
    for _ in range(500):
        n, labels, stats, _ = cv2.connectedComponentsWithStats(binary, 8)
        if n != ref_n or not np.array_equal(labels, ref_labels):
            with lock: errors.append(n)
        with lock: finished += 1

threads = [threading.Thread(target=worker) for _ in range(8)]
[t.start() for t in threads]
[t.join() for t in threads]
print(cv2.__version__, "finished:", finished, "errors:", len(errors))
# 4.10.0 finished: 4000 errors: 0  |  5.0.0 finished: 4000 errors: 0
```

That said, there are two things in the original snippet that could explain the symptom without OpenCV being involved. `pool.~ThreadPool()` calls the destructor explicitly, and the destructor runs a second time when `pool` goes out of scope — that is undefined behaviour in C++, and depending on how that ThreadPool handles shutdown it can produce this kind of intermittent symptom. Also, the `std::atomic` declaration appears commented out in the snippet; if the real counters were plain `int`, the `first++` / `last++` from several threads is a data race and "5 4" can simply be a lost increment.

I cannot test the original environment (3.3.1 / Windows / VS2015), so a platform-specific problem in that old version cannot be discarded.. but at least on current versions the function looks thread-safe under heavy concurrent load. Hope this helps!
